File: spring_security/eg05_sb3_oauth2_resource_server_jwt_symmetric_key/src/main/java/com/sc/sb/auth/config/ScUserDetail.java
package com.sc.sb.auth.config;

import com.sc.sb.auth.entity.ScUser;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;

import java.util.Collection;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

/**
 * DTO Object used by Spring Security. When authenticating in Spring Security uses this DTO to get users:
 * - Active,
 * - Roles/Authorities
 * - Password
 * - Any other custom data
 *
 * org.springframework.security.core.userdetails.UserDetails is used to create
 * org.springframework.security.core.Authentication
 *
 * Authentication is the main Object that is used by Spring Security's Security Context.
 *
 * Authentication Object can be DI. It contains current user's:
 * - Username
 * - is authenticated
 * - password
 * - Roles/Authorities
 */
public class ScUserDetail implements UserDetails {

    private final ScUser scUser;
    private final List<String> scRoles;

    public ScUserDetail(ScUser scUser, List<String> scRoles) {
        this.scUser = scUser;
        this.scRoles = scRoles;
    }

    @Override
    public Collection<? extends GrantedAuthority> getAuthorities() {
        return scRoles.stream()
                .filter(Objects::nonNull)
                .map(SimpleGrantedAuthority::new)
                .toList();
    }

    @Override
    public String getPassword() {
        return scUser.getUserPassword();
    }

    @Override
    public String getUsername() {
        return scUser.getUserName();
    }

    @Override
    public boolean isAccountNonExpired() {
        return true;
    }

    @Override
    public boolean isAccountNonLocked() {
        return true;
    }

    @Override
    public boolean isCredentialsNonExpired() {
        return true;
    }

    @Override
    public boolean isEnabled() {
        return true;
    }
}

File: spring_security/eg06_kotlin_sb3_oauth2_resource_server_jwt_symmetric_key/src/main/kotlin/com/sc/sb/auth/config/ScUserDetail.kt
package com.sc.sb.auth.config

import com.sc.sb.auth.entity.ScUser
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.UserDetails

class ScUserDetail(
    private val scUser: ScUser,
    private val scRoles: Collection<String>
) : UserDetails {
    override fun getAuthorities(): Collection<GrantedAuthority> = scRoles
        .filter { it.isNotBlank() }
        .map { SimpleGrantedAuthority(it) }

    override fun getPassword(): String = scUser.userPassword
    override fun getUsername(): String = scUser.userName
    override fun isAccountNonExpired(): Boolean = true
    override fun isAccountNonLocked(): Boolean = true
    override fun isCredentialsNonExpired(): Boolean = true
    override fun isEnabled(): Boolean = true
}